e2e/cmd/keys.go

```diff
@@ -0,0 +1,252 @@
+				return errors.Wrap(err, "write keys toml")
+			}
+
+			return nil
+		},
+	}
+
+	return cmd
+}
+
+// verifyKeyNodeType checks if the node exists in the manifest and if the key type is allowed for the node.
+func verifyKeyNodeType(def app.Definition, cfg key.UploadConfig) error {
+	if err := cfg.Type.Verify(); err != nil {
+		return err
+	}
+
+	if cfg.Type != key.EOA {
+		// Non-EOA keys must be not already defined in the manifest.
+		_, ok := def.Manifest.Keys[cfg.Name][cfg.Type]
+		if ok {
+			return errors.New("key already exists in manifest")
+		}
+	}
+
+	if cfg.Type == key.EOA {
+		eoaRole := eoa.Role(cfg.Name)
+		if err := eoaRole.Verify(); err != nil {
+			return errors.Wrap(err, "verifying name as eoa type")
+		}
+
+		account, ok := eoa.AccountForRole(def.Testnet.Network, eoaRole)
+		if !ok {
+			return errors.New("eoa account not found")
+		}
+
+		if account.Type != eoa.TypeSecret {
+			return errors.New("cannot create eoa key for non secret account")
+		}
+
+		if account.Address != (common.Address{}) {
+			return errors.New("cannot create eoa key already defined", "addr", account.Address.Hex())
+		}
+
+		return nil
+	}
+
+	for _, node := range def.Testnet.Nodes {
+		if node.Name == cfg.Name {
+			if cfg.Type == key.P2PExecution {
+				return errors.New("cannot create execution key for halo node")
+			}
+
+			return nil
+		}
+	}
+
+	for _, evm := range def.Testnet.OmniEVMs {
+		if evm.InstanceName == cfg.Name {
+			if cfg.Type != key.P2PExecution {
+				return errors.New("only execution keys allowed for evm nodes")
+			}
+
+			return nil
+		}
+	}
+
+	return errors.New("node not found")
+}
+
+// allKeyConfigs returns all key configurations for a given manifest.
+func allKeyConfigs(def app.Definition) []key.UploadConfig {
+	var cfgs []key.UploadConfig
+
+	// We need a relayer and monitor EOA key
+	cfgs = append(cfgs,
+		key.UploadConfig{
+			Network: def.Testnet.Network,
+			Name:    string(eoa.RoleRelayer),
+			Type:    key.EOA,
+		}, key.UploadConfig{
+			Network: def.Testnet.Network,
+			Name:    string(eoa.RoleMonitor),
+			Type:    key.EOA,
+		},
+	)
+
+	// All evm's need execution P2P keys
+	for _, evm := range def.Testnet.OmniEVMs {
+		cfgs = append(cfgs, key.UploadConfig{
+			Network: def.Testnet.Network,
+			Name:    evm.InstanceName,
+			Type:    key.P2PExecution,
+		})
+	}
+
+	for _, node := range def.Testnet.Nodes {
+		// All halo's need consensus P2P keys
+		cfgs = append(cfgs, key.UploadConfig{
+			Network: def.Testnet.Network,
+			Name:    node.Name,
+			Type:    key.P2PConsensus,
+		})
+
+		// Validators need validator keys
+		if node.Mode == types.ModeValidator {
+			cfgs = append(cfgs, key.UploadConfig{
+				Network: def.Testnet.Network,
+				Name:    node.Name,
+				Type:    key.Validator,
+			})
+		}
+	}
+
+	return cfgs
+}
+
+// printKeysToml prints the keys to as toml.
+func printKeysToml(keys map[string]map[key.Type]string, w io.Writer) error {
+	keysOnly := struct {
```

Comment:
should this write empty map as well? else do a check and return

Reply:
this isn't production code, exhaustive validation not super critical